File: src/test/python/advent2023/day1/day1.py
import re

values = []
with open("../../../resources/year2023/day1/input") as file:
    for line in file:
        values.append(line.rstrip())

replacements_dictionary = {
    'one': 'on1e',
    'two': 'tw2o',
    'three': 'thr3e',
    'four': 'fo4ur',
    'five': 'fi5ve',
    'six': 'si6x',
    'seven': 'sev7en',
    'eight': 'ei8ght',
    'nine': 'ni9ne'
}


def keep_only_digits(full_line):
    return re.sub('\\D', '', full_line)


def replace_literal_digits_with_digits(line_with_literals):
    lines_without_literals = line_with_literals
    for key, value in replacements_dictionary.items():
        lines_without_literals = lines_without_literals.replace(key, value)
    return lines_without_literals


def extract_first_and_last_digit(digits_only_value):
    return int(digits_only_value[0] + digits_only_value[-1])


# day 1 ex 1
print("day 1 ex 1")
summed = 0
for curr_line in values:
    digits_only = keep_only_digits(curr_line)
    summed += extract_first_and_last_digit(digits_only)
print("result : " + str(summed))
assert summed == 54877

# day 1 ex 2
print("day 1 ex 2")
summed = 0
for curr_line in values:
    curr_line = replace_literal_digits_with_digits(curr_line)
    digits_only = keep_only_digits(curr_line)
    summed += extract_first_and_last_digit(digits_only)
print("result : " + str(summed))
assert summed == 54100

File: src/main/java/org/likid/aoc/year2023/day1/Day1.java
package org.likid.aoc.year2023.day1;

import org.likid.aoc.util.AbstractDay;

import java.util.List;

public class Day1 extends AbstractDay<Long, Long> {

    public Day1(List<String> data) {
        super(data);
    }

    private static String replaceLiteralDigitsWithDigits(String line) {
        return line.replaceAll("nine", "ni9ne")
                .replaceAll("eight", "eig8ht")
                .replaceAll("seven", "sev7en")
                .replaceAll("six", "si6x")
                .replaceAll("five", "fi5ve")
                .replaceAll("four", "fo4ur")
                .replaceAll("three", "thr3ee")
                .replaceAll("two", "tw2o")
                .replaceAll("one", "on1e");
    }

    private static String keepOnlyDigits(String line) {
        return line.replaceAll("\\D", "");
    }

    private static Long extractFirstAndLastDigit(String val) {
        String[] split = val.split("");
        return Long.valueOf(getString(split, 0) + getString(split, split.length - 1));
    }

    private static String getString(String[] split, int index) {
        String val = split[index];
        return "".equals(val) ? "0" : val;
    }

    @Override
    public Long ex1() {
        return data.stream()
                .map(Day1::keepOnlyDigits)
                .mapToLong(Day1::extractFirstAndLastDigit)
                .sum();
    }

    @Override
    public Long ex2() {
        return data.stream()
                .map(Day1::replaceLiteralDigitsWithDigits)
                .map(Day1::keepOnlyDigits)
                .mapToLong(Day1::extractFirstAndLastDigit)
                .sum();
    }
}
